synapseclient/core/retry.py

```diff
@@ -38,9 +38,45 @@
 ]
 
 
-def with_retry(function, verbose=False,
-               retry_status_codes=[429, 500, 502, 503, 504], expected_status_codes=[],
-               retry_errors=[], retry_exceptions=[],
+def with_retry_network(function, verbose=False,
+                       retry_status_codes=[429, 500, 502, 503, 504], expected_status_codes=[], retry_errors=[],
+                       retry_exceptions=[], retries=3, wait=1, back_off=2, max_wait=30):
+    logger = _get_logger(verbose)
+
+    def _handle_status_code(response):
+        retry = False
+        # Check if we got a retry-able error
+        if response is None or not hasattr(response, 'status_code'):
+            return retry
+        if (
+            (expected_status_codes and response.status_code not in expected_status_codes) or
+            (retry_status_codes and response.status_code in retry_status_codes)
+        ):
+            response_message = _get_message(response)
+            retry = True
+            logger.debug("retrying on status code: %s" % str(response.status_code))
+            # TODO: this was originally printed regardless of 'verbose' was that behavior correct?
+            logger.debug(str(response_message))
+            # if (response.status_code == 429) and (wait > 10):
+            #     logger.warning('%s...\n' % response_message)
+            #     logger.warning('Retrying in %i seconds' % wait)
+        elif response.status_code not in range(200, 299):
+            # For all other non 200 messages look for retryable errors in the body or reason field
+            response_message = _get_message(response)
+            if any([msg.lower() in response_message.lower() for msg in retry_errors]):
+                retry = True
+                logger.debug('retrying %s' % response_message)
+            # special case for message throttling
+            elif 'Please slow down.  You may send a maximum of 10 message' in response:
+                logger.debug("retrying " + response_message)
+                raise SynapseThrottling(response_message, wait=16)
+        return retry
+    return with_retry(function, _handle_status_code, verbose=verbose,
+                      retry_errors=retry_errors, retry_exceptions=retry_exceptions, retries=retries, wait=wait,
+                      back_off=back_off, max_wait=max_wait)
+
+
+def with_retry(function, retry_evaluator=None, verbose=False, retry_errors=[], retry_exceptions=[],
```

Comment:
Can we also use this to generalize the retry logic for these cases with a custom evaluator function?

https://github.com/Sage-Bionetworks/synapsePythonClient/blob/develop/synapseclient/client.py#L3126